proposed/flexible-http.md

```diff
@@ -0,0 +1,101 @@
+
+This design doc describes a set of flexible, high-performance APIs for HTTP.
+
+## Target customers
+
+This API is intended for advanced users who need control or performance beyond what `HttpClient` offers, and to eventually replace the `SocketsHttpHandler` implementation.
+
+## APIs
+
+The APIs are split into three groups:
+
+- Low-level APIs
+- Compositional APIs
+- High-level APIs
+
+## Low-level API
+
+Low-level APIs implement HTTP without any additional features or opinions. They return HTTP data direct to the user without transformation, and are built to support new HTTP features (e.g. HTTP/2 and HTTP/3 extension frames) in the future.
+
+This API consists of an abstraction and concrete implementations:
+
+- `HttpConnection`, the abstraction.
+- `Http1Connection`, `Http2Connection`, `Http3Connection`.
+
+Performance is of high priority:
+
+- Operate on `Span<byte>` rather than `string` and `Uri`.
+    - Validation will only check for protocol correctness. Things like ASCII or URI validation will not be performed.
+- Zero allocations (steady-state).
+
+Ease of use is an explicit non-goal:
+
+- We assume users will know more about HTTP than `HttpClient` currently requires of them.
+- It is okay if it takes more code to use than `HttpClient`.
+
+These APIs are constructed on top of `Stream`. This has some implications worth noting:
+
+- Connections are not established by these APIs: the user must establish the connection.
+- Connecting via a proxy, or over TLS is not done by this API: the user must pass in a `Stream` that already has these established.
+
+"Prepared" headers will enable dynamic table compression and pre-validation/serialization equivalent to existing internal "known header" concept of `SocketsHttpHandler`.
+
+Huffman compression will be opt-in on a per-header basis.
+
+The API must be efficient for forwarding scenarios, as in reverse proxies.
+
+## Compositional APIs
+
+Compositional APIs are built on top of low-level APIs to provide additional features, as well as some opinionated implemenations.
+
+This will mean a `HttpConnection` implementation implementing some features:
+
+- Automatic decompression.
+- Automatic redirects.
+- Connection pooling.
+- Version negotiation.
+    - ALPN, Upgrade header, ALT-SVC.
+- Authentication.
+
+As well as some additional APIs for working with the `HttpConnection` API:
+
+- Header parsing.
+- Proxy implemenations
+    - HTTP proxy as `HttpConnection`.
+    - CONNECT proxy as `Stream`.
+    - SOCKS proxy as `Stream`.
+- Wrapper to work with `HttpConnection` as a `Stream`.
+- JSON helper extensions on top of `HttpConnection`.
+
+## High-level APIs
+
+This is a `HttpMessageHandler` implementation for use with `HttpClient`. It will either wrap a user-supplied `HttpConnection`, or will provide `SocketsHttpHandler`-parity functionality.
+
+Once we are confident in implementation and features, this would replace `SocketsHttpHandler`.
+
+Expect 100 Continue will be implemented here.
```

Comment:
I'd have expected 100-continue to be a lower level feature. I guess that depends on how you plan to pass body data into a connection. In todays model where you pass it an HttpContent, the connection has to own the 100-continue timing and copy the content to the network. If instead the caller directly passed bytes to be sent, it could control its own 100-continue timing.

Reply:
Current plan is that the low level API will provide informational responses; it's up to the caller (be it customer or higher level API) to act on it.

Reply:
> If instead the caller directly passed bytes to be sent, it could control its own 100-continue timing.

Sorry meant to reply to this part too -- this is exactly how content is handled.

Reply:
I could imagine 100-continue handling as a compositional API.

Reply:
I'll move it there, that makes sense to try.